File: 2024/Day2/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

func timer(name string) func() {
	start := time.Now()

	return func() {
		fmt.Printf("%s took %v\n", name, time.Since(start))
	}
}

func remove(slice []int, i int) []int {
	newSlice := make([]int, 0, len(slice)-1)
	newSlice = append(newSlice, slice[:i]...)
	return append(newSlice, slice[i+1:]...)
}

func abs(x int) int {
	if x < 0 {
		return -x
	}

	return x
}

func sign(x int) int {
	if x < 0 {
		return -1
	}

	return 1
}

func partOne(lists [][]int) int {
	safe := 0
outer:
	for _, list := range lists {
		if increasing := list[1] - list[0]; increasing != 0 && abs(increasing) >= 1 && abs(increasing) <= 3 {
			increasing = sign(increasing)
			for i := range list[1 : len(list)-1] {
				if curDiff := list[i+2] - list[i+1]; curDiff != 0 && sign(curDiff) == increasing {
					if !(abs(curDiff) >= 1 && abs(curDiff) <= 3) {
						continue outer
					}
				} else {
					continue outer
				}

			}

			safe++
		}
	}

	return safe
}

func partTwo(lists [][]int) int {
	safe := 0
	for _, list := range lists {
		if partOne([][]int{list}) == 1 {
			safe++
		} else {
			for i := range list {
				if partOne([][]int{remove(list, i)}) == 1 {
					safe++
					break
				}
			}
		}
	}

	return safe
}

func parseInput() [][]int {
	file, err := os.ReadFile("input.txt")
	if err != nil {
		_, _ = fmt.Fprintln(os.Stderr, "File open error: ", err)
		return nil
	}

	lines := strings.Split(strings.TrimSpace(string(file)), "\n")
	res := make([][]int, 0, len(lines))
	for _, line := range lines {
		nums := strings.Fields(line)

		list := make([]int, 0, len(nums))
		for _, num := range nums {
			n, _ := strconv.Atoi(num)
			list = append(list, n)
		}

		res = append(res, list)
	}

	return res
}

func main() {
	input := parseInput()
	if input == nil {
		return
	}

	// Part One
	func() {
		defer timer("Part One")()
		one := partOne(input)
		println("Part One:", one)
	}()

	// Part Two
	func() {
		defer timer("Part Two")()
		two := partTwo(input)
		println("Part Two:", two)
	}()
}

File: 2024/Day2/src/main.rs
use std::{fs, io, time::Instant};

fn timer(name: &str) -> impl FnOnce() {
    let start = Instant::now();
    let name = name.to_string();

    move || {
        let duration = Instant::now() - start;
        println!("{} took {:?}", name, duration);
    }
}

fn parse_input() -> Result<Vec<Vec<i32>>, io::Error> {
    let file = match fs::read_to_string("input.txt") {
        Ok(content) => content,
        Err(err) => {
            eprintln!("File open error: {err}");
            return Err(err);
        }
    };

    Ok(file
        .trim()
        .lines()
        .map(|x| x.split_whitespace())
        .map(|x| {
            x.map(|y| str::parse::<i32>(y).unwrap())
                .collect::<Vec<i32>>()
        })
        .collect::<Vec<Vec<i32>>>())
}

fn part_one<T>(lists: &[T]) -> i32
where
    T: AsRef<[i32]>,
{
    lists
        .iter()
        .map(|x| {
            let x = x.as_ref();
            let mut increasing = x[1] - x[0];
            if increasing.abs().clamp(1, 3) == increasing.abs() {
                increasing = increasing.signum();

                (x.len() - 2
                    == (1..x.len() - 1)
                        .map(|i| x[i + 1] - x[i])
                        .take_while(|&v| {
                            v != 0 && v.abs().clamp(1, 3) == v.abs() && v.signum() == increasing
                        })
                        .count()) as i32
            } else {
                0
            }
        })
        .sum()
}

fn part_two(lists: &[Vec<i32>]) -> i32 {
    lists
        .iter()
        .map(|x| {
            if part_one(&[&x]) == 1
                || (0..x.len())
                    .map(|i| {
                        let mut v = x.clone();
                        v.remove(i);
                        v
                    })
                    .any(|v| part_one(&[v]) == 1)
            {
                1
            } else {
                0
            }
        })
        .sum()
}

fn main() {
    let input = match parse_input() {
        Ok(parsed) => parsed,
        Err(_) => return,
    };

    let timer_one = timer("Part One");
    let one = part_one(&input);
    timer_one();
    println!("Part One: {one}");

    let timer_two = timer("Part Two");
    let two = part_two(&input);
    timer_two();
    println!("Part Two: {two}");
}
